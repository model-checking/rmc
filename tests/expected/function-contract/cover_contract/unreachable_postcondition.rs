// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

// Test `cover_contract` functionality, which fails verification for a unsatisfiable precondition
// or unreachable postcondition.
// See https://github.com/model-checking/kani/issues/2793

// Test that verification fails for an unreachable postcondition.

// The precondition is satisfiable, but the postcondition is unreachable because the code panics.
// Test that in the case where the postcondition check's property status is unreachable (because the function panics)
// we change the status to failure.
#[kani::requires(a > 5)]
#[kani::ensures(|result: &u32| *result == a)]
fn unreachable_postcondition(a: u32) -> u32 {
    panic!("panic")
}

#[kani::proof_for_contract(unreachable_postcondition)]
fn prove_unreachable_postcondition() {
    let x: u32 = kani::any();
    unreachable_postcondition(x);
}

// Unreachable postcondition because the function never returns.
// Test that in the case where the postcondition check's property status is undetermined because of unwinding failures,
// we change the status to failure.
#[kani::ensures(|result: &u32| *result == 7)]
fn never_return() -> u32 {
    loop {}
    7
}

#[kani::proof_for_contract(never_return)]
#[kani::unwind(5)]
fn prove_never_return() {
    never_return();
}
