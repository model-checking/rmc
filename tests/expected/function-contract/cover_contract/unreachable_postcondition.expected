
contract_cover.1\
	 - Status: FAILURE\
	 - Description: "The contract's postcondition is reachable."

unwind.0\
	 - Status: FAILURE\
	 - Description: "unwinding assertion loop 0"

VERIFICATION:- FAILED

assertion.1\
	 - Status: FAILURE\
	 - Description: "panic"

contract_cover.1\
	 - Status: SUCCESS\
	 - Description: "The contract's precondition is satisfiable."

contract_cover.2\
	 - Status: FAILURE\
	 - Description: "The contract's postcondition is reachable."

VERIFICATION:- FAILED
